File: backend/routes/auth_routes.py
from flask import Blueprint, jsonify

from controllers import auth_controller
from middleware.auth import require_auth
from middleware.validation import validate_request
from schemas.auth_schemas import (
    LoginSchema,
    LogoutSchema,
    RefreshTokenSchema,
    RegisterSchema,
    SendOtpSchema,
    SetPasswordSchema,
    VerifyOtpSchema,
    validate_login,
    validate_logout,
    validate_refresh,
    validate_register,
    validate_send_otp,
    validate_set_password,
    validate_verify_otp,
)

auth_bp = Blueprint("auth", __name__)


@auth_bp.post("/send-otp")
@validate_request(validate_send_otp)
def send_otp_route(data: SendOtpSchema):
    body, status = auth_controller.send_otp(data)
    return jsonify(body), status


@auth_bp.post("/verify-otp")
@validate_request(validate_verify_otp)
def verify_otp_route(data: VerifyOtpSchema):
    body, status = auth_controller.verify_otp(data)
    return jsonify(body), status


@auth_bp.post("/register")
@validate_request(validate_register)
def register_route(data: RegisterSchema):
    body, status = auth_controller.register(data)
    return jsonify(body), status


@auth_bp.post("/login")
@validate_request(validate_login)
def login_route(data: LoginSchema):
    body, status = auth_controller.login(data)
    return jsonify(body), status


@auth_bp.post("/logout")
@require_auth
@validate_request(validate_logout)
def logout_route(data: LogoutSchema):
    body, status = auth_controller.logout(data)
    return jsonify(body), status


@auth_bp.get("/validate-invite-token/<token>")
def validate_invite_token_route(token: str):
    body, status = auth_controller.validate_invite_token(token)
    return jsonify(body), status


@auth_bp.post("/set-password")
@validate_request(validate_set_password)
def set_password_route(data: SetPasswordSchema):
    body, status = auth_controller.set_password(data)
    return jsonify(body), status


@auth_bp.post("/refresh")
@validate_request(validate_refresh)
def refresh_route(data: RefreshTokenSchema):
    body, status = auth_controller.refresh(data)
    return jsonify(body), status

File: backend/routes/authRoutes.js
const express = require('express');
const authController = require('../controllers/authController');
const {
  loginRateLimiter,
  sendOtpRateLimiter,
  verifyOtpRateLimiter
} = require('../middleware/rateLimiter');
const { requireAuth } = require('../middleware/auth');
const {
  validateRequest,
  sendOtpSchema,
  verifyOtpSchema,
  registerSchema,
  loginSchema,
  logoutSchema,
  setPasswordSchema,
  refreshSchema,
  changePasswordSchema
} = require('../middleware/validation');

const router = express.Router();

// 1. Send OTP (Phone Verification dispatch)
router.post('/send-otp', sendOtpRateLimiter, validateRequest(sendOtpSchema), authController.sendOtp);

// 2. Verify OTP
router.post('/verify-otp', verifyOtpRateLimiter, validateRequest(verifyOtpSchema), authController.verifyOtp);

// 3. Register (Donor or Hospital Admin)
router.post('/register', validateRequest(registerSchema), authController.register);

// 4. Login (Email/Password or Phone/OTP)
router.post('/login', loginRateLimiter, validateRequest(loginSchema), authController.login);

// 5. Logout
router.post('/logout', validateRequest(logoutSchema), authController.logout);

// 6. Validate Invite Token (For staff invitation verification)
router.get('/validate-invite-token/:token', authController.validateInviteToken);

// 7. Set Password (Invited staff password activation)
router.post('/set-password', validateRequest(setPasswordSchema), authController.setPassword);

// 8. Refresh Token
router.post('/refresh', validateRequest(refreshSchema), authController.refresh);

// 9. Change Password (For direct staff password rotation)
router.post('/change-password', requireAuth, validateRequest(changePasswordSchema), authController.changePassword);

module.exports = router;
